Sources/Network/Network+URLSessionResource.swift

```diff
@@ -0,0 +1,103 @@
+import Foundation
+
+extension Network {
+
+    public struct URLSessionResource {
+
+        public let baseRequestMaking: BaseRequestMaking<URLRequest>
+        public let errorDecoding: ErrorDecoding<Data, URLResponse>
+
+        public let interceptors: [URLSessionResourceInterceptor]
+
+        public internal(set) var retryState: Retry.State
+        public let retryActionPriority: Retry.Action.CompareClosure
+
+        public init(
+            baseRequestMaking: BaseRequestMaking<URLRequest>,
```

Comment:
Just to confirm my understanding on the use of protocol witnesses here:

Instead of having multiple `Resource` protocols, each for their purpose (data decoding, error decoding...), we now inject protocol witnesses.

These protocol witnesses are concrete generic types that make that desired functionality available.

By doing this we are able to remove generic constraint complexity from our protocols and concrete types.

Am I making sense or missing something? 😅

Reply:
That's more or less it, yes 👍 

In essence, each protocol is converted to a witness that models/encapsulates the same behavior, which as you said greatly reduces generic requirements complexity when composed (vs protocol composition) without losing the actual abstraction. 

Additionally, as witnesses are concrete types, they give us some nice things "for free":

- you can have N witnesses for the same base type, whereas normally a type can only conform once to a protocol.

- you can create static helpers to instantiate default witness for a particular scenario (e.g. JSON decoding, or for your own Endpoint type).

- you can manipulate witnesses like any other type, because they are concrete types (e.g. add a `map`, `flatMap`, `contraMap`, ...)

Hope this helps